### ISSUE TITLE
Unpredictable order for marshalled maps

### ISSUE THREAD

**Issue by gustafj** [go-yaml/yaml#444]
The problem stated in #194 is not yet solved (as mentioned by @HeavyWombat in his comment https://github.com/go-yaml/yaml/issues/194#issuecomment-440818819).

Running his test on the current version creates unpredictable output, e.g.:
```
properties:
  cc:
    quota_definitions:
      q256MB:
        non_basic_services_allowed: true
      q2GB:
        non_basic_services_allowed: true
      q4GB:
        non_basic_services_allowed: true
```
vs
```
properties:
  cc:
    quota_definitions:
      q4GB:
        non_basic_services_allowed: true
      q256MB:
        non_basic_services_allowed: true
      q2GB:
        non_basic_services_allowed: true
```

What is the reason for using a custom (and error prone) implementation of string comparison in: https://github.com/go-yaml/yaml/blob/7b8349ac747c6a24702b762d2c4fd9266cf4f1d6/sorter.go#L39-L40 ?

Any particular reason why lexicographical ordering is unwanted / not sufficient as is the result of a simple `return a.String() < b.String()`?

_I opened a new issue as there seems to be no action in the closed one and to highlight the issue, hope that is ok_

**Comment by niemeyer** [comment]
> Any particular reason why lexicographical ordering is unwanted / not sufficient as is the result of a simple return a.String() < b.String()?

The reason is yaml files are used to convey human-tailored data, which is supposed to be pleasant to read instead of just stable.  Our brains like when they see _10GB_ after _2GB_, and _version-2.10_, after _version-2.2_. We also know _0033_ is closer to _32_ than to _1_. That's why we don't just do the obvious.

But what you've found is a bug. I've fixed it in cdc409d in v3. I'll backport it to v2 later.

**Comment by gustafj** [comment]
That's a good reason and makes perfect sense!
Thanks for the fast fix 👍

**Comment by mkmik** [comment]
> But what you've found is a bug. I've fixed it in cdc409d in v3. I'll backport it to v2 later.

It appears that the backport hasn't been done yet (I can reproduce it in v2.4.0, see https://github.com/mkmik/yamlsortbug))

**Comment by mkmik** [comment]
PR with cherry-pick from v3 to v2 in #736
